articles/setting-up-own-login-vs-login-via-external-services/index.md

```diff
@@ -0,0 +1,68 @@
+    alt: Setting Up Own Login vs Login via External Services image
+
+---
+Before we delve into the comparative study on various methods of authenticating the users; it is necessary to answer one primary question: Why does a website, app or service need to have login flow? Is it impossible to cater to the userbase without it? With User Experience playing a pivotal role in the retention of users, the login flow needs to be seamless and intuitive. At the same time, data privacy and security cannot be compromised. With so many factors playing a role in the login and sign-up component of the service, which might seem as a minor component initially, turns into a significant component.
+<!--more-->
+
+### Need for Authentication
+* Any solution today on the internet today consists of interactive content, like commenting on articles or booking an appointment. All these interactions do one of four primary CRUD operations on a database(Create, Read, Update, and Delete). Out of these, except read-operations, should be authenticated at the minimum (You may do it otherwise if the model of service is based on full anonymity. Even in such cases, you should incorporate rate-limiting or other methods to prevent abuse). Read-operations also should be authenticated to some extent, for example, if a user orders something online, only the concerned vendor should be able to see the user's mobile number and address, whereas viewing this article does not need authentication.
+* Suppose you get 10k comments on one of the posts on your service, or you get 1k orders within hours of launch on your product on an e-commerce website. These events are updating the database, which, if done without authentication, can lead to one or more of these.
+* Unable to map the comments/orders to the respective user who did it.
+* Partial or complete Denial of Service (DoS). These updates could have been done by a malicious script which in case of limited quantity products in e-commerce(and similar) services render them not be in a state to cater to relevant users.
+
+![Unless secured,the database can be compromised](/engineering-education/setting-up-own-login-vs-login-via-external-services/hacker-stealing-representation.jpg)
+
+### Basic Setup for Login-Signup Flow
+ Login-Signup is the first line of security for any service. If this is breached, a malicious user can access data that is otherwise not allowed. The login interface should also verify for user access level during the process- vendors, delivery boy, buyer, etc. in case of e-commerce or editor, author, a viewer in case of a blog, and similarly for other services. At the bare minimum, the authentication should be on username/email with the password, which can be secured even further with multi-factor authentication. To ensure privacy and protection of the users, passwords must be hashed at all stages(even before sendingit to the server via API).
```

Comment:
```suggestion
 Login-Signup is the first line of security for any service. If this is breached, a malicious user can access data that is otherwise not allowed. The login interface should also verify for user access level during the process- vendors, delivery boy, buyer, etc. in case of e-commerce or editor, author, a viewer in case of a blog, and similarly for other services. At the bare minimum, the authentication should be on username/email with the password, which can be secured even further with multi-factor authentication. To ensure privacy and protection of the users, passwords must be hashed wherever they are stored, and encrypted whenever they are transmitted.
```
Transmitting a hashed password from the client to the server can be detrimental or at least redundant.

If the server stores the password hashed via the same method that the client uses, then any hashing and salting of stored passwords is moot if an attacker compromises the database - they can transmit the values from the database directly over the API to authenticate.